lapack/gonum/dlagv2.go

```diff
@@ -0,0 +1,209 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dlagv2 computes the Generalized Schur factorization of a real 2-by-2
+// matrix pencil (A,B) where B is upper triangular. This routine
+// computes orthogonal (rotation) matrices given by CSL, SNL and CSR,
+// SNR such that
+//
+// 1) if the pencil (A,B) has two real eigenvalues (include 0/0 or 1/0 types), then
+//  [ a11 a12 ] := [  CSL  SNL ] [ a11 a12 ] [  CSR -SNR ]
+//  [  0  a22 ]    [ -SNL  CSL ] [ a21 a22 ] [  SNR  CSR ]
+//
+//  [ b11 b12 ] := [  CSL  SNL ] [ b11 b12 ] [  CSR -SNR ]
+//  [  0  b22 ]    [ -SNL  CSL ] [  0  b22 ] [  SNR  CSR ],
+// 2) if the pencil (A,B) has a pair of complex conjugate eigenvalues, then
+//  [ a11 a12 ] := [  CSL  SNL ] [ a11 a12 ] [  CSR -SNR ]
+//  [ a21 a22 ]    [ -SNL  CSL ] [ a21 a22 ] [  SNR  CSR ]
+//
+//  [ b11  0  ] := [  CSL  SNL ] [ b11 b12 ] [  CSR -SNR ]
+//  [  0  b22 ]    [ -SNL  CSL ] [  0  b22 ] [  SNR  CSR ]
+// where b11 >= b22 > 0.
+//
+// On exit, A is overwritten by the A-part of the generalized Schur form
+// and B is overwritten by the B-part of the generalized Schur form.
+//
+// (alphar(k)+imag*alphai(k))/beta(k) are the eigenvalues of the
+// pencil (A,B), k=0,1 where imag = sqrt(-1). Note that beta(k) may
+// be zero. They are solely outputs and will be overwritten.
+//
+// Returned floats are cosines (cs) and sines (sn) of left (l) and right (r)
+// rotation matrices.
+//
+// Dlagv2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlagv2(a []float64, lda int, b []float64, ldb int, alphar, alphai, beta []float64) (csl, snl, csr, snr float64) {
+	const safmin = dlamchS
+	const ulp = dlamchP
+
+	// Shorthand float64 routines.
+	var (
+		max = math.Max
+		abs = math.Abs
+	)
+	// Scale A.
+	anorm := max(abs(a[0*lda+0])+abs(a[1*lda+0]),
+		abs(a[0*lda+1])+abs(a[1*lda+1]))
+	anorm = max(anorm, safmin)
+	ascale := 1. / anorm
```

Comment:
`1` since `anorm` is `float64` the untyped constant is interpreted in a float context.